File: app/src/main/java/com/example/homestay/ui/admin/AdminBookingAdapter.java
package com.example.homestay.ui.admin;

import android.content.res.ColorStateList;
import android.view.*;
import android.widget.*;
import androidx.recyclerview.widget.RecyclerView;
import com.example.homestay.R;
import com.example.homestay.data.model.*;
import com.example.homestay.utils.DisplayFormatter;
import com.google.android.material.button.MaterialButton;
import java.text.*;
import java.util.*;

public class AdminBookingAdapter extends RecyclerView.Adapter<AdminBookingAdapter.Holder> {
  public interface Listener {
    void onBooking(AdminBookingData booking);
  }

  private final List<AdminBookingData> items;
  private final Listener change, delete, details;

  public AdminBookingAdapter(
      List<AdminBookingData> items, Listener change, Listener delete, Listener details) {
    this.items = items;
    this.change = change;
    this.delete = delete;
    this.details = details;
  }

  public void updateBookings(List<AdminBookingData> value) {
    items.clear();
    items.addAll(value);
    notifyDataSetChanged();
  }

  public void removeBooking(String id) {
    for (int i = 0; i < items.size(); i++)
      if (items.get(i).getId().equals(id)) {
        items.remove(i);
        notifyItemRemoved(i);
        break;
      }
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  @Override
  public Holder onCreateViewHolder(ViewGroup p, int t) {
    return new Holder(
        LayoutInflater.from(p.getContext()).inflate(R.layout.item_admin_booking, p, false));
  }

  @Override
  public void onBindViewHolder(Holder h, int p) {
    h.bind(items.get(p));
  }

  public class Holder extends RecyclerView.ViewHolder {
    final TextView badge, id, room, user, dates, guests, price, payment, contact, meta;
    final MaterialButton statusButton;
    final ImageButton deleteButton;

    Holder(View v) {
      super(v);
      badge = v.findViewById(R.id.tv_status_badge);
      id = v.findViewById(R.id.tv_booking_id);
      room = v.findViewById(R.id.tv_room_name);
      user = v.findViewById(R.id.tv_user_name);
      dates = v.findViewById(R.id.tv_dates);
      guests = v.findViewById(R.id.tv_guest_count);
      price = v.findViewById(R.id.tv_total_price);
      payment = v.findViewById(R.id.tv_payment_method);
      contact = v.findViewById(R.id.tv_customer_contact);
      meta = v.findViewById(R.id.tv_booking_meta);
      statusButton = v.findViewById(R.id.btn_change_status);
      deleteButton = v.findViewById(R.id.btn_delete);
    }

    void bind(AdminBookingData b) {
      itemView.setOnClickListener(v -> details.onBooking(b));
      badge.setText(statusLabel(b.getStatus()));
      badge.setBackgroundTintList(ColorStateList.valueOf(statusColor(b.getStatus())));
      id.setText(DisplayFormatter.bookingCode(b.getId(), b.getCreatedAt()));
      room.setText(b.getRoom() == null ? "N/A" : b.getRoom().getName());
      user.setText(
          "Khách hàng: " + (b.getUser() == null ? "Chưa xác định" : b.getUser().getFullName()));
      dates.setText(date(b.getCheckInDate()) + " – " + date(b.getCheckOutDate()));
      guests.setText(b.getGuestCount() + " khách");
      price.setText(DisplayFormatter.vnd(b.getTotalPrice()));
      String method =
          "pay_on_site".equals(b.getPaymentMethod())
              ? "Thanh toán khi nhận phòng"
              : "qr_code".equals(b.getPaymentMethod()) ? "Thanh toán QR" : "Chưa thanh toán";
      payment.setText(method);
      contact.setText(
          (b.getUser() == null ? "Chưa có email" : b.getUser().getEmail())
              + " • "
              + (b.getUser() == null ? "Chưa có số điện thoại" : b.getUser().getPhone()));
      long nights =
          Math.max(
              1,
              com.example.homestay.domain.BookingCalculator.nights(
                  b.getCheckInDate(), b.getCheckOutDate()));
      String created =
          new SimpleDateFormat("dd/MM/yyyy HH:mm", new Locale("vi", "VN"))
              .format(new Date(b.getCreatedAt()));
      meta.setText(
          nights
              + " đêm • "
              + (b.getSlotId() == null ? "Cả phòng" : "Slot " + b.getSlotId())
              + " • Tạo lúc "
              + created);
      statusButton.setOnClickListener(v -> change.onBooking(b));
      deleteButton.setOnClickListener(v -> delete.onBooking(b));
    }
  }

  private static String date(long t) {
    return new SimpleDateFormat("dd/MM/yyyy HH:mm", new Locale("vi", "VN")).format(new Date(t));
  }

  private static String statusLabel(String s) {
    switch (s.toLowerCase()) {
      case "pending":
        return "Chờ duyệt";
      case "confirmed":
        return "Đã xác nhận";
      case "completed":
        return "Hoàn thành";
      case "cancelled":
        return "Đã hủy";
      default:
        return s;
    }
  }

  private static int statusColor(String s) {
    switch (s.toLowerCase()) {
      case "pending":
        return 0xffff9800;
      case "confirmed":
        return 0xff4caf50;
      case "cancelled":
        return 0xfff44336;
      case "completed":
        return 0xff2196f3;
      default:
        return 0xff9e9e9e;
    }
  }
}

File: app/src/main/java/com/example/homestay/ui/admin/AdminBookingAdapter.kt
package com.example.homestay.ui.admin

import android.graphics.drawable.ColorDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.example.homestay.R
import com.example.homestay.data.model.AdminBookingData
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class AdminBookingAdapter(
    private val bookings: MutableList<AdminBookingData>,
    private val onChangeStatusClick: (AdminBookingData) -> Unit,
    private val onDeleteClick: (AdminBookingData) -> Unit,
    private val onDetailsClick: (AdminBookingData) -> Unit
) : RecyclerView.Adapter<AdminBookingAdapter.BookingViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookingViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_admin_booking, parent, false)
        return BookingViewHolder(view)
    }

    override fun onBindViewHolder(holder: BookingViewHolder, position: Int) {
        val booking = bookings[position]
        holder.bind(booking)
    }

    override fun getItemCount() = bookings.size

    fun updateBookings(newBookings: List<AdminBookingData>) {
        bookings.clear()
        bookings.addAll(newBookings)
        notifyDataSetChanged()
    }

    fun removeBooking(bookingId: String) {
        val index = bookings.indexOfFirst { it.id == bookingId }
        if (index != -1) {
            bookings.removeAt(index)
            notifyItemRemoved(index)
        }
    }

    inner class BookingViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvStatusBadge: TextView = itemView.findViewById(R.id.tv_status_badge)
        private val tvBookingId: TextView = itemView.findViewById(R.id.tv_booking_id)
        private val tvRoomName: TextView = itemView.findViewById(R.id.tv_room_name)
        private val tvUserName: TextView = itemView.findViewById(R.id.tv_user_name)
        private val tvDates: TextView = itemView.findViewById(R.id.tv_dates)
        private val tvGuestCount: TextView = itemView.findViewById(R.id.tv_guest_count)
        private val tvTotalPrice: TextView = itemView.findViewById(R.id.tv_total_price)
        private val tvPaymentMethod: TextView = itemView.findViewById(R.id.tv_payment_method)
        private val tvCustomerContact: TextView = itemView.findViewById(R.id.tv_customer_contact)
        private val tvBookingMeta: TextView = itemView.findViewById(R.id.tv_booking_meta)
        private val btnChangeStatus: com.google.android.material.button.MaterialButton = itemView.findViewById(R.id.btn_change_status)
        private val btnDelete: ImageButton = itemView.findViewById(R.id.btn_delete)

        fun bind(booking: AdminBookingData) {
            itemView.setOnClickListener { onDetailsClick(booking) }
            // Status badge
            tvStatusBadge.text = booking.status.uppercase()
            val color = getStatusColor(booking.status)
            tvStatusBadge.background = ColorDrawable(color)
            
            // Booking ID (short)
            tvBookingId.text = "ID: ${booking.id.takeLast(8)}"
            
            // Room name
            tvRoomName.text = booking.room?.name ?: "N/A"
            
            // User name
            tvUserName.text = booking.user?.fullName ?: "N/A"
            
            // Dates
            val checkIn = formatDate(booking.checkInDate)
            val checkOut = formatDate(booking.checkOutDate)
            tvDates.text = "$checkIn - $checkOut"
            
            // Guest count
            tvGuestCount.text = "👥 ${booking.guestCount} người"
            
            // Total price
            tvTotalPrice.text = formatPrice(booking.totalPrice)
            
            // Payment method
            val paymentMethod = when (booking.paymentMethod) {
                "pay_on_site" -> "Thanh toán khi nhận phòng"
                "qr_code" -> "Thanh toán QR"
                else -> "Chưa thanh toán"
            }
            tvPaymentMethod.text = "💳 $paymentMethod"
            tvCustomerContact.text = "✉ ${booking.user?.email ?: "N/A"}  •  ☎ ${booking.user?.phone ?: "N/A"}"
            val nights = ((booking.checkOutDate - booking.checkInDate) / 86_400_000L).coerceAtLeast(1)
            val created = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale("vi", "VN")).format(Date(booking.createdAt))
            tvBookingMeta.text = "$nights đêm • Slot: ${booking.slotId ?: "Cả phòng"} • Tạo lúc $created"
            
            // Actions
            btnChangeStatus.setOnClickListener {
                onChangeStatusClick(booking)
            }
            
            btnDelete.setOnClickListener {
                onDeleteClick(booking)
            }
        }

        private fun formatDate(timestamp: Long): String {
            val sdf = SimpleDateFormat("dd/MM/yyyy", Locale("vi", "VN"))
            return sdf.format(Date(timestamp))
        }

        private fun formatPrice(price: Double): String {
            val formatter = NumberFormat.getNumberInstance(Locale("vi", "VN"))
            return "${formatter.format(price)} VNĐ"
        }

        private fun getStatusColor(status: String): Int {
            return when (status.lowercase()) {
                "pending" -> 0xFFFF9800.toInt() // Orange
                "confirmed" -> 0xFF4CAF50.toInt() // Green
                "cancelled" -> 0xFFF44336.toInt() // Red
                "completed" -> 0xFF2196F3.toInt() // Blue
                else -> 0xFF9E9E9E.toInt() // Gray
            }
        }
    }
}

